source/Octopus.Tentacle.Tests.Integration/Util/Builders/Decorators/Proxies/ServiceProxy.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System;
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+
+namespace Octopus.Tentacle.Tests.Integration.Util.Builders.Decorators.Proxies
+{
+    public abstract class ServiceProxy : DispatchProxyAsync
+    {
+        protected object? TargetService { get; private set; }
+
+        protected void SetTargetService(object service)
+        {
+            TargetService = service;
+        }
+
+        public override object Invoke(MethodInfo method, object[] args)
+        {
+            EnsureTargetServiceNotNull();
+            try
+            {
+                OnStartingInvocation(method);
+
+                return method.Invoke(TargetService, args)!;
+            }
+            catch (TargetInvocationException e)
+            {
+                OnInvocationException(method, e.InnerException!);
+
+                //we need to unwrap the TargetInvocationException
+                ExceptionDispatchInfo.Capture(e.InnerException).Throw();
```

Comment:
This throws the inner exception with a stack trace that doesn't get restarted from here (as it would be if we did `throw e.InnerException`)